launcher/container_runner.go

```diff
@@ -194,43 +211,114 @@ func (r *ContainerRunner) measureContainerClaims(ctx context.Context) error {
 		}
 	}
 
-	contianerSpec, err := r.container.Spec(ctx)
+	containerSpec, err := r.container.Spec(ctx)
 	if err != nil {
 		return err
 	}
-	for _, arg := range contianerSpec.Process.Args {
+	for _, arg := range containerSpec.Process.Args {
 		if err := r.attestAgent.MeasureEvent(cel.CosTlv{EventType: cel.ArgType, EventContent: []byte(arg)}); err != nil {
 			return err
 		}
 	}
-	for _, env := range contianerSpec.Process.Env {
+	for _, env := range containerSpec.Process.Env {
 		if err := r.attestAgent.MeasureEvent(cel.CosTlv{EventType: cel.EnvVarType, EventContent: []byte(env)}); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-// Run the container
-// Doesn't support container restart yet
-// Container output will always be redirected to stdio for now
-func (r *ContainerRunner) Run(ctx context.Context) error {
-	if err := r.measureContainerClaims(ctx); err != nil {
-		return fmt.Errorf("failed to measure container claims: %v", err)
+func getTTL(token []byte) (time.Duration, error) {
+	claims := &jwt.RegisteredClaims{}
+	_, _, err := new(jwt.Parser).ParseUnverified(string(token), claims)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse token: %w", err)
 	}
 
-	// Create hostTokenPath directory. Should come before NewTask.
+	now := time.Now()
+	if !now.Before(claims.ExpiresAt.Time) {
+		return 0, errors.New("token is expired")
+	}
+
+	return time.Until(claims.ExpiresAt.Time), nil
+}
+
+func (r *ContainerRunner) refreshToken(ctx context.Context) (time.Duration, error) {
 	token, err := r.attestAgent.Attest(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to retrieve attestation service token: %v", err)
+		return 0, fmt.Errorf("failed to retrieve attestation service token: %v", err)
+	}
+
+	filepath := path.Join(HostTokenPath, attestationVerifierTokenFile)
+	if err = os.WriteFile(filepath, token, 0644); err != nil {
+		return 0, fmt.Errorf("failed to write token to container mount source point: %v", err)
+	}
+
+	ttl, err := getTTL(token)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get token TTL: %v", err)
+	}
+
+	return time.Duration(float64(ttl) * r.refresher.refreshMultiplier), nil
+}
+
+func (r *ContainerRunner) fetchAndWriteToken(ctx context.Context, refreshMultiplier float64) error {
```

Comment:
I think this function can be simplified if we don't store the refresher in the `ContainerRunner`.
```go
func (r *ContainerRunner) fetchAndWriteToken(ctx context.Context) error {
	if err := os.MkdirAll(HostTokenPath, 0744); err != nil {
		return err
	}

	dur, err := r.refreshToken(ctx)
	if err != nil {
		return err
	}
	
	timer := time.NewTimer(dur)
	go func() {
		for {
			select {
			case <-ctx.Done():
				log.Printf("token refreshing stopped: %v", ctx.Err())
				return
			case <-timer.C:
				dur, err := r.refreshToken(ctx)
				if err != nil {
					log.Printf("failed to refresh attestation service token: %v", err)
					return
				}
				timer.Reset(dur)
			}
		}
	}()
	return nil
}
```

We then also shouldn't need a `done` channel as one is already built into the `ctx`.

Reply:
It looks like if we use `ctx.Done()`, we have to make sure the context can be cancelled (context.Background() is never cancelled by default so Done() will never be closed in that case).

This can be worked around by using `ctx, cancel := context.WithCancel(context.Background())`, then calling `cancel()` during cleanup, but not sure if this defeats the purpose of removing a dedicated `done` channel in the first place.

Reply:
Good catch, that's correct, we need to cancel the context once the container stops. Using `WithCancel` seems correct here as that will handle cancelling outstanding RPCs as well as cancelling the timer.

Putting something like:
```go
ctx, cancel := context.WithCanel(ctx)
defer cancel()
```
at the start of `Run` should work.